src/hotspot/share/nmt/regionsTree.hpp

```diff
@@ -0,0 +1,171 @@
+      inline bool is_released_begin() { return out_state() == VMATree::StateType::Released; }
+      inline bool is_reserved_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Reserved) == 1; }
+      inline VMATree::StateType in_state() { return _node->val().in.type(); }
+      inline VMATree::StateType out_state() { return _node->val().out.type(); }
+      inline size_t distance_from(NodeHelper& other) { return position() - other.position(); }
+      inline NativeCallStackStorage::StackIndex out_stack_index() { return _node->val().out.stack(); }
+      inline MemTag in_tag() { return _node->val().in.mem_tag(); }
+      inline MemTag out_tag() { return _node->val().out.mem_tag(); }
+      inline void set_in_tag(MemTag tag) { _node->val().in.set_tag(tag); }
+      inline void set_out_tag(MemTag tag) { _node->val().out.set_tag(tag); }
+      inline void print_on(outputStream* st) {
+        auto st_str = [&](int s){
+          return s == (int)VMATree::StateType::Released ? "Rl" :
+                 s ==  (int)VMATree::StateType::Reserved ? "Rv" : "Cm";
+        };
+        st->print_cr("pos: " INTPTR_FORMAT " "
+                     "%s, %s <|> %s, %s",
+                     p2i((address)position()),
+                     st_str((int)in_state()),
+                     NMTUtil::tag_to_name(in_tag()),
+                     st_str((int)out_state()),
+                     NMTUtil::tag_to_name(out_tag())
+                     );
+      }
+    };
+
+  void print_on(outputStream* st) {
+    visit_in_order([&](Node* node) {
+      NodeHelper curr(node);
+      curr.print_on(st);
+      return true;
+    });
+  }
+
+  template<typename F>
+  void visit_committed_regions(position start, size_t size, F func) {
+    size_t end = start + size + 1;
+    size_t comm_size = 0;
+    size_t base = start;
+
+    NodeHelper prev;
+    visit_range_in_order(start, end, [&](Node* node) {
+      NodeHelper curr(node);
+      if (prev.is_valid()) {
+        base = prev.position();
+        if (prev.is_committed_begin()) {
+          comm_size += curr.distance_from(prev);
+          if (!curr.is_committed_begin()) {
+            auto st = stack(curr);
+            CommittedMemoryRegion cmr((address)base, comm_size, st);
+            comm_size = 0;
+            if (!func(cmr))
+              return false;
+          }
+        }
+      }
+      prev = curr;
+      return true;
+    });
+  }
+
+  template<typename F>
+  void visit_reserved_regions(F func) {
+    NodeHelper begin_node, prev;
+    size_t rgn_size = 0;
+    size_t comm_size = 0;
+    size_t base = 0;
+
+    visit_in_order([&](Node* node) {
+      NodeHelper curr(node);
+      if (prev.is_valid()) {
+        rgn_size += curr.distance_from(prev);
+      } else {
+        begin_node = curr;
+        rgn_size = 0;
+      }
+      prev = curr;
+      if (curr.is_released_begin() || begin_node.out_tag() != curr.out_tag()) {
```

Comment:
```
inline bool is_released_begin() { return out_state() == VMATree::StateType::Released; }
```
Released means neither reserved nor committed.
In case two different regions are adjacent with different `MemTag`, the `out_tag()` comparison will catch it.